regionurlprovider.go

```diff
@@ -0,0 +1,159 @@
+		return "", errors.New("no regions available")
+	}
+
+	var bestRegionURL string
+	minAttempts := -1
+	for _, region := range hostnameSettings.regionSettings.Regions {
+		if minAttempts == -1 || hostnameSettings.regionURLAttempts[region.Url] < minAttempts {
+			minAttempts = hostnameSettings.regionURLAttempts[region.Url]
+			bestRegionURL = region.Url
+		}
+	}
+	hostnameSettings.regionURLAttempts[bestRegionURL]++
+
+	return bestRegionURL, nil
+}
+
+// ReportAttemptFailure reports a failed attempt to connect to a region, so that it can be avoided in future connection attempts
+func (r *regionURLProvider) ReportAttemptFailure(cloudHostname, regionURL string) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	hostnameSettings := r.hostnameSettingsCache[cloudHostname]
+	if hostnameSettings == nil {
+		return errors.New("ReportAttemptFailure: serverURL not found in cache")
+	}
+
+	// remove failed region from regionSettings
+	for i, region := range hostnameSettings.regionSettings.Regions {
+		if region.Url == regionURL {
+			hostnameSettings.regionSettings.Regions = append(
+				hostnameSettings.regionSettings.Regions[:i],
+				hostnameSettings.regionSettings.Regions[i+1:]...,
+			)
+			break
+		}
+	}
+
+	return nil
+}
+
+// assume this is being called within a lock
+func (r *regionURLProvider) refreshRegionSettings(cloudHostname, token string) error {
+	settingsURL := "https://" + cloudHostname + "/settings/regions"
+	req, err := http.NewRequest("GET", settingsURL, nil)
+	if err != nil {
+		return errors.New("refreshRegionSettings failed to create request: " + err.Error())
+	}
+	req.Header = http.Header{
+		"Authorization": []string{"Bearer " + token},
+	}
+
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return errors.New("refreshRegionSettings failed to fetch region settings. http status: " + resp.Status)
+	}
+
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return errors.New("refreshRegionSettings failed to read response body: " + err.Error())
+	}
+	regions := &livekit.RegionSettings{}
+	if err := protojson.Unmarshal(respBody, regions); err != nil {
+		return errors.New("refreshRegionSettings failed to decode region settings: " + err.Error())
+	}
+
+	r.hostnameSettingsCache[cloudHostname] = &hostnameSettingsCacheItem{
+		regionSettings:    regions,
+		updatedAt:         time.Now(),
+		regionURLAttempts: map[string]int{},
+	}
+
+	return nil
+}
+
+func parseCloudURL(serverURL string) (string, error) {
+	parsedURL, err := url.Parse(serverURL)
+	if err != nil {
+		return "", fmt.Errorf("invalid server url (%s): %v", serverURL, err)
+	}
+
+	if !isCloud(parsedURL.Hostname()) {
+		return "", errors.New("not a cloud url")
+	}
+
+	return parsedURL.Hostname(), nil
+}
+
+func isCloud(hostname string) bool {
+	return strings.HasSuffix(hostname, "livekit.cloud") || strings.HasSuffix(hostname, "livekit.io")
```

Comment:
Would be good to make this a config so that we can add/remove domains easily without changing code. Can also address the private cloud case. If config is empty, can return `false`.

Reply:
could we add the regions endpoint to oss so the request returns a url from config or formatted from the http host header? pattern based detection in the client feels fragile. if we want oss clients to return an error the endpoint could send back a non 200 response with a meaningful error message.

Reply:
I like the idea of avoiding code path changes based on string pattern matching. But I'll leave that for a subsequent PR as the JS SDK currently does this, and we still have more to discuss on this topic.